curve25519-dalek/src/backend/serial/u64/scalar.rs

```diff
@@ -164,34 +178,406 @@ impl Scalar52 {
             // VERIFICATION NOTE: Result is canonical
             to_nat(&s.limbs) < group_order(),
     {
-        assume(false);  // TODO: complete the proof
+        let ghost wide_input = bytes_wide_to_nat(bytes);
+
+        // Stage 1 assumption: the byte-to-word packing yields the expected little-endian value.
         let mut words = [0u64;8];
-        for i in 0..8 {
-            for j in 0..8 {
-                assume(false);
+        proof {
+            assert forall|k: int| 0 <= k < 8 implies words@[k] == 0 by {
+                assert(0 <= k < 8);
+                assert(#[trigger] words@[k] == 0);
+            };
+        }
+        for i in 0..8
+            invariant
+                0 <= i <= 8,
+                forall|k: int| 0 <= k < i ==> words@[k] as nat == word_from_bytes(bytes, k),
+                words_from_bytes_to_nat(bytes, i as int) + bytes_wide_to_nat_rec(
+                    bytes,
+                    (i as int) * 8,
+                ) == bytes_wide_to_nat(bytes),
+                forall|k: int| i <= k < 8 ==> words@[k] == 0,
+        {
+            proof {
+                assert(1u64 << 0 == 1u64) by (bit_vector);
+                assert(words@[i as int] == 0);
+            }
+            for j in 0..8
+                invariant
+                    0 <= j <= 8,
+                    i < 8,
+                    forall|k: int| 0 <= k < i ==> words@[k] as nat == word_from_bytes(bytes, k),
+                    forall|k: int| i < k < 8 ==> words@[k] == 0,
```

Comment:
This should be redundant, since it's not about `j` and is still within the `i`-loop